src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Windows.cs

```diff
@@ -0,0 +1,336 @@
+            //
+            // Since contentType is only CERT, singleContext should either be valid, or the
+            // function should have returned false.
+            if (actualType != ContentType || singleContext.IsInvalid)
+            {
+                singleContext.Dispose();
+                throw new CryptographicException();
+            }
+
+            CertificatePal pal = new CertificatePal(singleContext, deleteKeyContainer: false);
+            return pal;
+        }
+
+        private static SafeCertStoreHandle ImportPfx(
+            ref BagState bagState,
+            ReadOnlySpan<char> password,
+            X509KeyStorageFlags keyStorageFlags)
+        {
+            ArraySegment<byte> reassembled = bagState.ToPfx(password);
+            SafeCertStoreHandle storeHandle = ImportPfx(reassembled, password, keyStorageFlags);
+            CryptoPool.Return(reassembled);
+
+            return storeHandle;
+        }
+
+        private static unsafe SafeCertStoreHandle ImportPfx(
+            ReadOnlySpan<byte> data,
+            ReadOnlySpan<char> password,
+            X509KeyStorageFlags keyStorageFlags)
+        {
+            const int MaxStackPasswordLength = 64;
+            void* alloc = default;
+            Span<char> szPassword = stackalloc char[MaxStackPasswordLength + 1];
+            ReadOnlySpan<char> effectivePassword = szPassword;
+            Interop.Crypt32.PfxCertStoreFlags flags = MapKeyStorageFlags(keyStorageFlags);
+
+            try
+            {
+                if (password.Contains('\0'))
+                {
+                    effectivePassword = password;
+                }
+                else
+                {
+                    if (password.Length >= MaxStackPasswordLength)
+                    {
+                        alloc = NativeMemory.Alloc((uint)password.Length + 1, sizeof(char));
+                        szPassword = new Span<char>(alloc, password.Length + 1);
+                    }
+
+                    password.CopyTo(szPassword);
+                    szPassword[password.Length] = '\0';
+                }
+
+                SafeCertStoreHandle storeHandle;
+
+                fixed (byte* dataPtr = data)
+                fixed (char* szPtr = szPassword)
+                {
+                    Interop.Crypt32.DATA_BLOB blob = new((IntPtr)dataPtr, (uint)data.Length);
+
+                    storeHandle = Interop.Crypt32.PFXImportCertStore(
+                        ref blob,
+                        szPtr,
+                        flags);
+                }
+
+                if (storeHandle.IsInvalid)
+                {
+                    Exception e = Marshal.GetHRForLastWin32Error().ToCryptographicException();
+                    storeHandle.Dispose();
+                    throw e;
+                }
+
+                return storeHandle;
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(MemoryMarshal.AsBytes(szPassword));
```

Comment:
Probably doesn't make much of a difference but you could slice `szPassword` to `password.Length` to avoid zeroing unwritten portions of the eagerly allocated stack buffer.